sqld/src/http/hrana_over_http.rs

```diff
@@ -0,0 +1,157 @@
+            .map(|result| RespBody { result })
+            .map_err(|err| match err.downcast::<batch::StmtError>() {
+                Ok(stmt_err) => anyhow!(ResponseError::Stmt(stmt_err)),
+                Err(err) => err,
+            })
+            .context("Could not execute statement")
+    })
+    .await
+}
+
+pub async fn handle_batch(
+    db_factory: Arc<dyn DbFactory>,
+    req: hyper::Request<hyper::Body>,
+) -> Result<hyper::Response<hyper::Body>> {
+    #[derive(Debug, Deserialize)]
+    struct ReqBody {
+        batch: batch::proto::Batch,
+    }
+
+    #[derive(Debug, Serialize)]
+    struct RespBody {
+        result: batch::proto::BatchResult,
+    }
+
+    handle_request(db_factory, req, |db, req_body: ReqBody| async move {
+        batch::execute_batch(&*db, &req_body.batch)
+            .await
+            .map(|result| RespBody { result })
+            .map_err(|err| match err.downcast::<batch::BatchError>() {
+                Ok(batch_err) => anyhow!(ResponseError::Batch(batch_err)),
+                Err(err) => err,
+            })
+            .context("Could not execute batch")
+    })
+    .await
+}
+
+async fn handle_request<ReqBody, RespBody, F, Fut>(
+    db_factory: Arc<dyn DbFactory>,
+    req: hyper::Request<hyper::Body>,
+    f: F,
+) -> Result<hyper::Response<hyper::Body>>
+where
+    ReqBody: DeserializeOwned,
+    RespBody: Serialize,
+    F: FnOnce(Arc<dyn Database>, ReqBody) -> Fut,
+    Fut: Future<Output = Result<RespBody>>,
+{
+    let res: Result<_> = async move {
+        let req_body = hyper::body::to_bytes(req.into_body()).await?;
+        let req_body = serde_json::from_slice(&req_body)
+            .map_err(|e| ResponseError::BadRequestBody { source: e })?;
+
+        let db = db_factory
+            .create()
+            .await
+            .context("Could not create a database connection")?;
+        let resp_body = f(db, req_body).await?;
+
+        Ok(json_response(hyper::StatusCode::OK, &resp_body))
+    }
+    .await;
+
+    Ok(match res {
+        Ok(resp) => resp,
+        Err(err) => error_response(err.downcast::<ResponseError>()?),
+    })
+}
+
+fn error_response(err: ResponseError) -> hyper::Response<hyper::Body> {
+    use batch::{BatchError, StmtError};
+    let status = match &err {
+        ResponseError::BadRequestBody { .. } => hyper::StatusCode::BAD_REQUEST,
+        ResponseError::Stmt(err) => match err {
+            StmtError::SqlParse { .. }
+            | StmtError::SqlNoStmt
+            | StmtError::SqlManyStmts
+            | StmtError::ArgsInvalid { .. }
+            | StmtError::SqlInputError { .. } => hyper::StatusCode::BAD_REQUEST,
+            StmtError::ArgsBothPositionalAndNamed => hyper::StatusCode::NOT_IMPLEMENTED,
+            StmtError::TransactionTimeout | StmtError::TransactionBusy => {
+                hyper::StatusCode::SERVICE_UNAVAILABLE
+            }
+            StmtError::SqliteError { .. } => hyper::StatusCode::INTERNAL_SERVER_ERROR,
```

Comment:
Uhm regarding my comment before: I think this is the culprit

Reply:
I was just surprised because it came up in the logs as an internal error, which leads me to believe that there is something broken, although everything is fine, and this is simply a user error

Reply:
not sure
let's start with that and work on the error story later